### ISSUE TITLE
Queries calling stored procedures broken with recent MariaDB/MySQL

### ISSUE THREAD

**Issue by XanC** [PowerDNS/pdns#9421]
- Program: Authoritative
 - Issue type: Bug report

### Short description
Stored procedures for PowerDNS MySQL backend queries fail with error "Could not bind parameters to mysql statement" when the database server is MariaDB >= 10.2 or (I believe but have not personally seen) MySQL >= 5.7.

### Environment
 - Operating system: Debian Stretch (oldstable), Buster (stable), Bullseye (testing)
 - Software version: 4.0.3 (patched by Debian to enable stored procedures), 4.1.6, 4.3.0.
 - Software source: Debian, compiled on the Debian system (eg, when trying 4.3.0 on Stretch, I compiled the 4.3.0 Debian pdns sources on Stretch)

### Steps to reproduce
1. Prepare a test machine running Debian.  Stretch is the most versatile: it'll start out working, because it uses MariaDB 10.1, and then you can upgrade the database and then see the error.  Buster and Bullseye use newer versions of PowerDNS and MariaDB.  The problem is the newer version of MariaDB, so the error should be present from the start on those.

2. Install pdns-backend-mysql and mariadb-server.  If asked to configure pdns-backend-mysql with dbconfig-common, select No.

3. Create a file (let's say called test.conf) in /etc/powerdns/pdns.d) containing:
```
query-logging=yes
launch=gmysql
gmysql-user=powerdns
gmysql-dbname=pdnstest
gmysql-password=ueoaueoa
```

(And delete any other files, such as bind.conf or pdns.local.gmysql.conf, that are in that directory.)

4. Run the following to set up the database:
```
echo "create database pdnstest" | mysql

echo "grant all on pdnstest.* to powerdns@localhost identified by 'ueoaueoa'" | mysql
```

If on Stretch, modify /usr/share/doc/pdns-backend-mysql/schema.mysql.sql and replace all instances of 255 with 127.  (Otherwise the indexes get too big.)
```
cat /usr/share/doc/pdns-backend-mysql/schema.mysql.sql | mysql pdnstest

echo "insert into domains(name, type) values('example.com', 'MASTER');" | mysql pdnstest

echo "insert into records(domain_id, name, type, content, ttl) values(1, 'example.com', 'SOA', 'localhost admin.example.com 1 10380 3600 604800 3600', 86400);" | mysql pdnstest

echo "insert into records(domain_id, name, type, content, ttl) values(1, 'example.com', 'A', '1.2.3.4', 3600);" | mysql pdnstest
```

5. Run (assuming the dnsutils package is installed): `dig @localhost example.com`

and observe that the first query performed (and showing in /var/log/syslog) is:
`SELECT content,ttl,prio,type,domain_id,disabled,name,auth FROM records WHERE disabled=0 and type=? and name=?`

This is the "basic" query.

6. Let's create a stored procedure which does nothing other than run the same query:

```
echo "DELIMITER //
CREATE PROCEDURE basic_query(incoming_type varchar(10), incoming_name varchar(255))
BEGIN
SELECT content,ttl,prio,type,domain_id,disabled,name,auth FROM records WHERE disabled=0 and type=incoming_type and name=incoming_name;
END//" | mysql pdnstest
```

7. Modify /etc/powerdns/pdns.d/test.conf to use our new query, by adding the line:
```
gmysql-basic-query=CALL basic_query(?, ?)
```

8. Restart PowerDNS: `service pdns restart`

9. Again run: `dig @localhost example.com`

And observe that we get a SERVFAIL in response, and that /var/log/syslog contains:
```
Backend reported permanent error which prevented lookup (GSQLBackend get: Could not bind parameters to mysql statement: CALL basic_query(?, ?): Prepared statement contains no metadata), aborting
```

That is, unless you did this on Stretch.  If you did it on Stretch, then this failure did not take place.  You can add this to your repository sources:
```
deb http://ftp.osuosl.org/pub/mariadb/repo/10.2/debian stretch main
```
and upgrade MariaDB to 10.2, and then (after restarting pdns) see that it now fails.

### Expected behaviour
The query should not fail regardless of the underlying database version.

### Actual behaviour
SERVFAIL as described above.

### Other information
There are two places in smysql.cc which can throw that same "could not bind parameters" error.  I discovered that it's in the second one.  It's this code:
```
#if MYSQL_VERSION_ID >= 50500
      if (d_residx >= d_resnum) {
        mysql_stmt_free_result(d_stmt);
        while(!mysql_stmt_next_result(d_stmt)) {
          if ((err = mysql_stmt_store_result(d_stmt))) {
            string error(mysql_stmt_error(d_stmt));
            releaseStatement();
            throw SSqlException("Could not store mysql statement while processing additional sets: " + d_query + string(": ") + error);
          }
          d_resnum = mysql_stmt_num_rows(d_stmt);
          // XXX: For some reason mysql_stmt_result_metadata returns NULL here, so we cannot
          // ensure row field count matches first result set.
          if (d_resnum > 0) { // ignore empty result set
            if (d_res_bind != nullptr && (err = mysql_stmt_bind_result(d_stmt, d_res_bind))) {
              string error(mysql_stmt_error(d_stmt));
              releaseStatement();
              throw SSqlException("Could not bind parameters to mysql statement: " + d_query + string(": ") + error);
            }
            d_residx = 0;
            break;
          }
          mysql_stmt_free_result(d_stmt);
        }
      }
#endif
```

Here's my mailing list post about it:
https://mailman.powerdns.com/pipermail/pdns-users/2020-August/026810.html

Thank you!

**Comment by XanC** [comment]
Thanks for helping format that properly!

**Comment by zeha** [comment]
Maybe relevant mysql doc:
https://dev.mysql.com/doc/c-api/8.0/en/c-api-prepared-call-statements.html

#9422 attempts to add a test using the simple suggested SP.

**Comment by rgacogne** [comment]
This issue should now be resolved in master since #9423 has been merged. @XanC, thanks again for reporting this! Would you by any chance be able to test the fix, either by testing our master builds (available on https://repo.powerdns.com/) or by applying the patch then compiling yourself if you can't use our packages?

**Comment by XanC** [comment]
At the moment my main goal is to get this working on Stretch with updated MariaDB, so I merged the fix into the Debian Stretch 4.0.3 sources.  It now works perfectly.

Thanks a million!

As far as I can tell, the only loose end (which is really just idle curiosity on my part) is why this worked on older MySQL/MariaDB and not on newer.  Called procedure support was originally added to PowerDNS in early 2018:
https://github.com/PowerDNS/pdns/pull/6134
which was apparently sufficient for MariaDB <= 10.1 and MySQL <= 5.6 but not for later versions.  Weird.

Regardless, it looks like there's now an official test for stored procedures, so things should be in great shape on this front going forward.

**Comment by rgacogne** [comment]
Alright, thank you for testing! The corresponding fix has been merged to master and is tagged for backporting to 4.3.x so I'll close this issue.
